### PR TITLE
Init trajectory

### PR DESCRIPTION
**Pull request by ogibson**
# Description

# Details
#### URL to issue
https://biomage.atlassian.net/browse/BIOMAGE-1946

#### Link to staging deployment URL (or set N/A)
https://ui-sara-worker275.scp-staging.biomage.net/

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [ ] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [x] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [ ] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [ ] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [x] Started end-to-end tests on the latest commit.

### Documentation updates
- [ ] Relevant Github READMEs updated **or** no GitHub README updates required.
- [ ] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

File: r/Dockerfile [hunk 1]
```diff
@@ -26,6 +26,12 @@ RUN R -e "renv::init(bare = TRUE); options(renv.consent = TRUE); renv::restore(l
 COPY renv.lock .
 RUN R -e "options(renv.consent = TRUE); renv::restore(lockfile='renv.lock')"
 
+RUN R -e "BiocManager::install(c('BiocGenerics', 'DelayedArray', 'DelayedMatrixStats','limma', 'S4Vectors', 'SingleCellExperiment','SummarizedExperiment', 'batchelor','Matrix.utils'))"
```
Comment (ogibson) [inline r/Dockerfile]: All these should be moved to renv.lock

File: r/Dockerfile [hunk 2]
```diff
@@ -61,4 +76,6 @@ RUN pip install -U jedi radian PyYAML watchdog[watchmedo]
 ADD R ./R
 COPY work.R .
 
-CMD watchmedo auto-restart --directory=. --pattern='*.R' --recursive -- Rscript work.R
\ No newline at end of file
+CMD Rscript work.R
+
+#CMD watchmedo auto-restart --directory=. --pattern='*.R' --recursive -- Rscript work.R
```
Comment (ogibson) [inline r/Dockerfile]:
```suggestion
CMD watchmedo auto-restart --directory=. --pattern='*.R' --recursive -- Rscript work.R
```

File: python/src/worker/tasks/factory.py [hunk 3]
```diff
@@ -31,6 +32,8 @@ class TaskFactory:
             GetDoubletScore,
             GetMitochondrialContent,
             MarkerHeatmap,
+            getPseudoTime,
+            getTrajectoryGraph,
```
Comment (kafkasl) [inline python/src/worker/tasks/factory.py]: make the casing consisten (first uppercase)

File: python/src/worker/tasks/trajectory_analysis.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import backoff
+import requests
+from aws_xray_sdk.core import xray_recorder
+
+from ..config import config
+from ..result import Result
+from ..tasks import Task
+
+
+class getPseudoTime(Task):
```
Comment (kafkasl) [inline python/src/worker/tasks/trajectory_analysis.py]: create a different file for each task

File: r/R/trajectory_analysis.R [hunk 5]
```diff
@@ -0,0 +1,68 @@
+runTrajectoryAnalysis <- function(req, data) {
+  root_nodes <- req$body$rootNodes
+
+  cell_data <- generateGraphData(data)
+  cell_data <- monocle3::order_cells(cell_data, reduction_method = "UMAP", root_pr_nodes = root_nodes)
+
+  # monocle3::plot_cells(cell_data,color_cells_by = "pseudotime",label_cell_groups=FALSE,label_leaves=FALSE,label_branch_points=FALSE,graph_label_size=1.5)
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R]: why is this commented out?

File: r/R/trajectory_analysis.R [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+  pseudotime <- as.data.frame(cell_data@principal_graph_aux@listData$UMAP$pseudotime)
+  pseudotime$cells_id <- data@meta.data$cells_id
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R]: nooooo why "cells_id"?! we should work to remove this, not add more 😩

File: r/R/trajectory_analysis.R [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+runGenerateTrajectoryGraph <- function(req, data) {
+  cell_data <- generateGraphData(data)
+  node_coords <- t(cell_data@principal_graph_aux[["UMAP"]]$dp_mst)
+  umap_coords <- as.data.frame(SingleCellExperiment::reducedDims(cell_data)[["UMAP"]])
+
+  # node coordinates
+  # get connected nodes
+  connected_nodes <- list()
+  for (node in rownames(node_coords)) {
+    connected_nodes[[node]] <- as.list(names(cell_data@principal_graph[["UMAP"]][[which(rownames(node_coords) == node)]][[1]]))
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R]: this could be made clearer, it's super difficult to read. for starters, extract the `which(rownames(node_coords) == node)` to an aptly named variable.

File: r/R/trajectory_analysis.R [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+  # convert list to json
+  node_umap_coords <- RJSONIO::toJSON(node_umap_coords)
+  return(node_umap_coords)
+}
+
+
+generateGraphData <- function(data) {
+  cell_data <- SeuratWrappers::as.cell_data_set(data)
+
+  set.seed(42)
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R]: we should extract the hardcoded seed to a variable stored in sysdata.R

Comment (gerbeldo) [reply r/R/trajectory_analysis.R]: and name it `ULTIMATE_SEED`, of course

File: r/renv/activate.R [hunk 9]
```diff
@@ -2,7 +2,7 @@
 local({
 
   # the requested version of renv
-  version <- "0.13.2"
```
Comment (gerbeldo) [inline r/renv/activate.R]: this should not be updated here

File: r/tests/testthat/test-trajectory_analysis.R [hunk 10]
```diff
@@ -0,0 +1,94 @@
+mock_scdata <- function(filt_cell_id = "") {
+  data("pbmc_small", package = "SeuratObject", envir = environment())
+  pbmc_small$cells_id <- 0:(ncol(pbmc_small) - 1)
+  if (all(filt_cell_id != "")) {
+    pbmc_small <- subset(pbmc_small, cells = names(pbmc_small$cells_id[which(!pbmc_small$cells_id %in% filt_cell_id)]))
```
Comment (gerbeldo) [inline r/tests/testthat/test-trajectory_analysis.R]: maybe make it readable. do not fear assignments

File: r/Dockerfile [hunk 11]
```diff
@@ -34,6 +40,15 @@ RUN R -e "install.packages('covr')"
 RUN R -e "install.packages('xml2')"
 
 
+RUN apt-get install -y libudunits2-dev
+RUN apt-get install -y libgdal-dev
+
+RUN R -e "install.packages('sf')"
+RUN R -e "devtools::install_github('cole-trapnell-lab/monocle3')"
```
Comment (saracastel) [inline r/Dockerfile]: why is this repeated 2 times? (see line 33)

Comment (gerbeldo) [reply r/Dockerfile]: all of this should go to the package.lock file. you can install things from github without devtools. using `renv::install('cole-trapnell-lab/monocle3')` should work. and then when snapshotting it adds github as the source.

File: r/Dockerfile [hunk 12]
```diff
@@ -26,6 +26,12 @@ RUN R -e "renv::init(bare = TRUE); options(renv.consent = TRUE); renv::restore(l
 COPY renv.lock .
 RUN R -e "options(renv.consent = TRUE); renv::restore(lockfile='renv.lock')"
 
+RUN R -e "BiocManager::install(c('BiocGenerics', 'DelayedArray', 'DelayedMatrixStats','limma', 'S4Vectors', 'SingleCellExperiment','SummarizedExperiment', 'batchelor','Matrix.utils'))"
+
+RUN R -e "install.packages('devtools')"
+RUN R -e "devtools::install_github('cole-trapnell-lab/leidenbase')"
+RUN R -e "devtools::install_github('cole-trapnell-lab/monocle3')"
```
Comment (ogibson) [inline r/Dockerfile]:
```suggestion
```

File: r/Dockerfile (hunk 11 above)
Comment (ogibson) [inline r/Dockerfile]:
```suggestion
RUN R -e "install.packages('devtools')"
RUN R -e "devtools::install_github('cole-trapnell-lab/leidenbase')"
RUN R -e "devtools::install_github('cole-trapnell-lab/monocle3')"
```

File: r/Dockerfile [hunk 13]
```diff
@@ -34,6 +37,17 @@ RUN R -e "install.packages('covr')"
...
 
+RUN apt-get install -y libudunits2-dev
+RUN apt-get install -y libgdal-dev
+
+RUN R -e "install.packages('sf')"
+RUN R -e "install.packages('devtools')"
+RUN R -e "devtools::install_github('cole-trapnell-lab/leidenbase')"
+RUN R -e "devtools::install_github('cole-trapnell-lab/monocle3')"
+RUN R -e "install.packages('R.utils')"
+RUN R -e "remotes::install_github('satijalab/seurat-wrappers')"
```
Comment (ogibson) [inline r/Dockerfile]:
```suggestion
```

File: r/Dockerfile [hunk 14]
```diff
@@ -26,6 +26,9 @@ RUN R -e "renv::init(bare = TRUE); options(renv.consent = TRUE); renv::restore(l
 COPY renv.lock .
 RUN R -e "options(renv.consent = TRUE); renv::restore(lockfile='renv.lock')"
```
Comment (ogibson) [inline r/Dockerfile]:
```suggestion
RUN apt-get install -y libudunits2-dev
RUN apt-get install -y libgdal-dev
RUN R -e "options(renv.consent = TRUE); renv::restore(lockfile='renv.lock')"
```

File: r/Dockerfile [hunk 15]
```diff
@@ -24,8 +24,13 @@ RUN R -e "renv::init(bare = TRUE); options(renv.consent = TRUE); renv::restore(l
 
 # use renv::snapshot() while R dependency updates are quick to build
 COPY renv.lock .
+RUN apt-get install -y libudunits2-dev
+RUN apt-get install -y libgdal-dev
```
Comment (ogibson) [inline r/Dockerfile]:
```suggestion
RUN apt-get install -y libgdal-dev
RUN apt-get -y install cmake
```

File: r/R/trajectory_analysis.R (hunk 5 above)
Comment (saracastel) [reply r/R/trajectory_analysis.R]: was there just for development. Removed

File: python/src/worker/tasks/trajectory_analysis.py (hunk 4 above)
Comment (saracastel) [reply python/src/worker/tasks/trajectory_analysis.py]: > create a different file for each task

why different files? Doesn't make sense to have both tasks in the same file, since they're two steps of the trajectory analysis?

File: r/Dockerfile [hunk 16]
```diff
@@ -5,6 +5,7 @@ WORKDIR /src
 # install required debian packages to make R packages work
 COPY setup ./setup
 COPY requirements_debian.txt .
+RUN R -e "print('')"
```
Comment (ogibson) [inline r/Dockerfile]:
```suggestion
```

File: python/src/worker/tasks/trajectory_analysis.py (hunk 4 above)
Comment (kafkasl) [reply python/src/worker/tasks/trajectory_analysis.py]: in many languages is a common practice to have only one class in each file, and to name the file with the class name. This makes it very easy to open files by name and see how many different tasks are there (just looking at the number of files and names).  Moreover, short files are easier to debug and test, specially in this case where the classes are independent (even if they are part of the same functionality)

**Review by kafkasl — CHANGES_REQUESTED** [review]
split the classes into files

File: r/data-raw/sysdata.R [hunk 17]
```diff
@@ -6,4 +6,6 @@ error_codes <- list(
   UNHANDLED_ERROR = "R_WORKER_UNHANDLED_ERROR"
 )
 
-usethis::use_data(error_codes, internal = TRUE, overwrite = TRUE)
+ULTIMATE_SEED <- 42
```
Comment (aerlaut) [inline r/data-raw/sysdata.R:9]: What is this `ULTIMATE_SEED`? What is it used for? Why ultimate? Is there not "penultimate"? Are we not going to change it? Is it a constant? Why 42? Why not 911?

Comment (gerbeldo) [reply r/data-raw/sysdata.R:9]: it's a random seed! and could be 911. could be just called seed, but that would be boring. it's ultimate because it is [the answer to the ultimate question of life the universe and everything](https://news.mit.edu/2019/answer-life-universe-and-everything-sum-three-cubes-mathematics-0910#:~:text=The%20number%2042%20is%20especially,the%20Universe%2C%20and%20Everything.%E2%80%9D).

File: r/Dockerfile [hunk 18]
```diff
@@ -71,6 +71,8 @@ RUN echo "python3-pip" >> sysdeps_run.txt && \
     cat sysdeps_run.txt | xargs ./install_debian_packages.sh && \
     rm -rf *
 
+
+
```
Comment (gerbeldo) [inline r/Dockerfile]:
```suggestion
```

File: python/src/worker/tasks/trajectory_graph.py [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+    def compute(self):
+        request = {}
+
+        r = requests.post(
+            f"{config.R_WORKER_URL}/v0/runGenerateTrajectoryGraph",
+            headers={"content-type": "application/json"},
+            data=json.dumps(request),
+        )
+
+        # raise an exception if an HTTPError if one occurred because otherwise r.json() will fail
```
Comment (gerbeldo) [inline python/src/worker/tasks/trajectory_graph.py]:
```suggestion
        # raise an exception if an HTTPError occurred. otherwise r.json() will fail
```

File: r/R/trajectory_analysis.R [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+  cell_data <- generateGraphData(data)
+  cell_data <- monocle3::order_cells(cell_data, reduction_method = "UMAP", root_pr_nodes = root_nodes)
+
+  pseudotime <- as.data.frame(cell_data@principal_graph_aux@listData$UMAP$pseudotime)
+  pseudotime$cells_id <- data@meta.data$cells_id
+  pseudotime <- pseudotime[order(pseudotime$cells_id), ]
+  pseudotime <- pseudotime %>%
+    tidyr::complete(cells_id = seq(0, max(data@meta.data$cells_id))) %>%
+    dplyr::select(-cells_id)
+  return(unname(pseudotime))
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R]:
```suggestion

  return(unname(pseudotime))
```

File: r/R/trajectory_analysis.R [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+  return(node_umap_coords)
+}
+
+
+generateGraphData <- function(data) {
+  cell_data <- SeuratWrappers::as.cell_data_set(data)
+
+  set.seed(ULTIMATE_SEED)
+
+  cell_data <- monocle3::cluster_cells(cds = cell_data, reduction_method = "UMAP")
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R:108]: Can't we use pre-calculated clustering to learn the graph? will we show the new clusters anywhere during trajectory analysis? because this function uses leiden clustering by default and this could confuse users, given that we only support louvain at the moment. The clusters will probably be different than what we show in qc/data-exp

File: r/R/trajectory_analysis.R [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+  root_nodes <- req$body$rootNodes
+
+  cell_data <- generateGraphData(data)
+  cell_data <- monocle3::order_cells(cell_data, reduction_method = "UMAP", root_pr_nodes = root_nodes)
+
+  pseudotime <- as.data.frame(cell_data@principal_graph_aux@listData$UMAP$pseudotime)
+  pseudotime$cells_id <- data@meta.data$cells_id
+  pseudotime <- pseudotime[order(pseudotime$cells_id), ]
+  pseudotime <- pseudotime %>%
+    tidyr::complete(cells_id = seq(0, max(data@meta.data$cells_id))) %>%
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R]: why is this necessary? not all cells get assigned a pseudo time?

Comment (gerbeldo) [reply r/R/trajectory_analysis.R]: or is it filtered cells?

Comment (ogibson) [reply r/R/trajectory_analysis.R]: Filtered cells

File: r/R/trajectory_analysis.R (hunk 21 above)
Comment (ogibson) [reply r/R/trajectory_analysis.R:108]: The user won't actually see the clusters. These are used for background calculation only, and the problem is that apparently from what I understand we can't really use our louvain clusters for this.

Comment (gerbeldo) [reply r/R/trajectory_analysis.R:108]: :(

File: r/data-raw/sysdata.R (hunk 17 above)
Comment (ogibson) [reply r/data-raw/sysdata.R:9]: A perfect answer

**Review by kafkasl — APPROVED** [review]

File: r/R/img_plot.R [hunk 23]
```diff
@@ -0,0 +1,174 @@
+runRidgePlot <- function(req,data,config){
```
Comment (gerbeldo) [inline r/R/img_plot.R]: what is this file?

File: r/Dockerfile [hunk 24]
```diff
@@ -16,6 +16,8 @@ RUN R -e "if(Sys.getenv('GITHUB_PAT') == '') stop('need to export GITHUB_PAT')"
 ENV RENV_LIB=/src/lib
 RUN echo ".libPaths(c('$RENV_LIB', .libPaths()))" >> $(R RHOME)/etc/Rprofile.site
 
+RUN apt-get update && apt-get install -y procps
```
Comment (gerbeldo) [inline r/Dockerfile]: this is installed, take a look at sysdeps_build_debian. After the upgrade R and general refactor, linux dependencies go there. And this one is present.

File: r/R/trajectory_analysis.R [hunk 25]
```diff
@@ -0,0 +1,68 @@
+runTrajectoryAnalysis <- function(req, data) {
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R]: Add roxygen docs to all functions.

File: r/R/trajectory_analysis.R [hunk 26]
```diff
@@ -0,0 +1,68 @@
...
+  pseudotime$cells_id <- data@meta.data$cells_id
+  pseudotime <- pseudotime[order(pseudotime$cells_id), ]
+  pseudotime <- pseudotime %>%
+    tidyr::complete(cells_id = seq(0, max(data@meta.data$cells_id))) %>%
+    dplyr::select(-cells_id)
+
+  return(unname(pseudotime))
+}
+
+runGenerateTrajectoryGraph <- function(req, data) {
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R:18]: why does this function take a req argument? it does not seem to be used

File: r/R/trajectory_analysis.R [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+    node_coords_list[[i]]["connected_nodes"] <- list(connected_nodes[[i]])
+  }
+
+  # umap data
+  # fill in the NULL values for filtered cells
+  umap_coords$cells_id <- data@meta.data$cells_id
+  umap_coords <- umap_coords[order(umap_coords$cells_id), ]
+  umap_coords <- umap_coords %>%
+    tidyr::complete(cells_id = seq(0, max(data@meta.data$cells_id))) %>%
+    dplyr::select(-cells_id)
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R]: all of this could be refactored, as it is used in the other function as well

File: r/Dockerfile (hunk 24 above)
Comment (gerbeldo) [reply r/Dockerfile]:
```suggestion
```

File: r/Dockerfile [hunk 28]
```diff
@@ -16,6 +16,7 @@ RUN R -e "if(Sys.getenv('GITHUB_PAT') == '') stop('need to export GITHUB_PAT')"
 ENV RENV_LIB=/src/lib
 RUN echo ".libPaths(c('$RENV_LIB', .libPaths()))" >> $(R RHOME)/etc/Rprofile.site
 
+
```
Comment (gerbeldo) [inline r/Dockerfile]:
```suggestion
```

File: r/R/trajectory_analysis.R (hunk 26 above)
Comment (saracastel) [reply r/R/trajectory_analysis.R:18]: it was not necessary. Removed

File: r/R/trajectory_analysis.R [hunk 29]
```diff
@@ -0,0 +1,82 @@
...
+  # node + umap data
+  node_umap_coords <- list(nodes = node_coords_list, umap = umap_coords_list)
+
+  # convert list to json
+  node_umap_coords <- RJSONIO::toJSON(node_umap_coords)
+  return(node_umap_coords)
+}
+
+
+runTrajectoryAnalysis <- function(req, data) {
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R:40]: rename to estimate pseudotime or something like that.

File: python/src/worker/tasks/trajectory_graph.py [hunk 30]
```diff
@@ -0,0 +1,37 @@
...
+
+        # Return a list of formatted results.
+        return [Result(raw_result), Result(raw_result)]
+
+    @xray_recorder.capture("GetTrajectoryGraph.compute")
+    @backoff.on_exception(
+        backoff.expo, requests.exceptions.RequestException, max_time=30
+    )
+    def compute(self):
+        request = {}
```
Comment (aerlaut) [inline python/src/worker/tasks/trajectory_graph.py:25]: Doesn't the worker need to take the embedding passed from the UI?

File: r/R/trajectory_analysis.R [hunk 31]
```diff
@@ -0,0 +1,127 @@
...
+#' Returns a json object with UMAP and nodes coordinates, to be used for the initial plot (embedding + trajectory nodes)
+#'
+#' This represents the first step of the trajectory analysis. It allows the creation of the initial trajectory analysis plot, which is composed of an embedding and starting nodes (vertices in the lines).
+#' This plot is used to select the root nodes among the starting nodes. The root nodes will be then used in the following step of the trajectory analysis for pseudotime calculation.
+#'
+#' @param data SeuratObject
+#'
+#' @return a json with nodes and umap coordinates
+#' @export
+runGenerateTrajectoryGraph <- function(data) {
```
Comment (aerlaut) [inline r/R/trajectory_analysis.R]: I think this should be `function(req, data)` because the parameters passed to it (see https://github.com/biomage-ltd/worker/blob/0de12af5a7abf8735c6944b38ae427d29df2eb6d/r/work.R#L57) is `(req, data)`

File: python/src/worker/tasks/trajectory_graph.py (hunk 30 above)
Comment (aerlaut) [reply python/src/worker/tasks/trajectory_graph.py:25]: Embedding is already contained inside the Seurat object, so we don't need to pass it in again.

File: r/R/trajectory_analysis.R (hunk 31 above)
Comment (gerbeldo) [reply r/R/trajectory_analysis.R]: that's a good catch!

File: r/R/trajectory_analysis.R [hunk 32]
```diff
@@ -0,0 +1,127 @@
+#' Generate UMAP and nodes coordinates for the initial trajectory analysis plot
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R]:
```suggestion
#' Generate UMAP and node coordinates for the initial trajectory analysis plot
```

File: r/R/trajectory_analysis.R [hunk 33]
```diff
@@ -0,0 +1,127 @@
...
+#'
+#' Returns a json object with UMAP and nodes coordinates, to be used for the initial plot (embedding + trajectory nodes)
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R]:
```suggestion
#' Returns a json object with UMAP and node coordinates, 
#' to be used by the UI for the initial plot (embedding + trajectory nodes)
```

File: r/R/trajectory_analysis.R [hunk 34]
```diff
@@ -0,0 +1,127 @@
...
+#' @export
+fillNullForFilteredCells <- function(df, data) {
+  # add cells_id column
+  df$cells_id <- data@meta.data$cells_id
+  # get max value for cells_id
+  max_value <- max(data@meta.data$cells_id)
+  # order by cells_id
+  df <- df[order(df$cells_id), ]
+  # add NULL values for filtered cells and remove cells_id column
+  df <- df %>%
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R]:
```suggestion
  df$cells_id <- data@meta.data$cells_id
  max_value <- max(data@meta.data$cells_id)
  df <- df[order(df$cells_id), ]
  
  df <- df %>%
```

Comment (gerbeldo) [reply r/R/trajectory_analysis.R]: These comments are not necessary. the code is pretty self explanatory! And is also clearer thanks to the function docs

File: r/R/trajectory_analysis.R [hunk 35]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+
+#' Fill in the NULL values for filtered cells
+#'
+#' When a dataframe with cell barcodes and some associated information is generated from a filtered Seurat object, it will not contain the cell barcodes that were filtered out.
+#' However, the UI needs the whole array of ordered and unfiltered cell ids.
+#' For this reason, there is a need to add to the dataframe the corresponding cell ids for cell barcodes that were filtered out and fill the values with NULL.
+#'
+#' @param df a dataframe with barcodes as rownames
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R]:
```suggestion
#' @param df data.frame with barcodes as rownames
```

File: r/R/trajectory_analysis.R [hunk 36]
```diff
@@ -0,0 +1,127 @@
...
+#' @param data Seurat object with cells id stored in: data@@meta.data$cells_id
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R]:
```suggestion
#' @param data Seurat object with cells_id
```

File: r/R/trajectory_analysis.R [hunk 37]
```diff
@@ -0,0 +1,127 @@
...
+  pseudotime <- fillNullForFilteredCells(pseudotime, data)
+
+  return(unname(pseudotime))
+}
+
+
+#' Convert Seurat object to Monocle3 cell_data_set object, cluster cells and learn graph
+#'
+#' In order to perform the trajectory analysis with Monocle3, the Seurat object needs to be converted to a Monocle3 cell_data_set object.
+#' This function calculates also embeddings and learn the graph.
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R]: does this function calculate the embedding? doesn't it use the embedding already present?

Comment (saracastel) [reply r/R/trajectory_analysis.R]: To my understanding, for the first iteration we need to calculate the embeddings for the cell_data_set object. With the second iteration the embeddings will be downloaded from S3 using the Etag

File: r/R/trajectory_analysis.R [hunk 38]
```diff
@@ -0,0 +1,137 @@
...
+}
+
+
+#' Calculate pseudotime
+#'
+#' Order the cells and generate an array of pseudotime values, based on
+#' the node ids of the root nodes.
+#'
+#' Pseudotime is a cell value that shows the relative progression
+#' of the cell in time for a given biological process.
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R]:
```suggestion
#' Pseudotime is a value that shows the relative progression
#' of a cell through a path in gene expression space.
```

File: r/R/trajectory_analysis.R [hunk 39]
```diff
@@ -0,0 +1,137 @@
...
+#' These values will be used to plot the cells along a continuous path
+#' that represents the evolution of the process.
+#' This plot represents the final plot of the trajectory analysis, which shows
+#' a map of how cells “differentiate” through time, starting from cells in the
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R]:
```suggestion
#' a map of how cells expression changes, starting from cells with smaller
```

File: r/R/trajectory_analysis.R [hunk 40]
```diff
@@ -0,0 +1,137 @@
...
+#' earlier pseudotime growing into cells in later pseudotime, following the trajectory.
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R]:
```suggestion
#' pseudotime values to cells with larger pseudotimes, along a trajectory.
```

File: r/R/trajectory_analysis.R (hunk 37 above)
Comment (gerbeldo) [reply r/R/trajectory_analysis.R]: [monocle3::learn_graph](https://rdrr.io/github/cole-trapnell-lab/monocle3/man/learn_graph.html) does not calculate an embedding (check the docs. embeddings are calculated using [monocle3::reduce_dimensions](https://rdrr.io/github/cole-trapnell-lab/monocle3/man/reduce_dimension.html)). [SeuratWrappers::as.cell_data_set](https://rdrr.io/github/satijalab/seurat-wrappers/man/as.cell_data_set.html#:~:text=Cell%20emebeddings%20are%20transferred%20over%20to%20the%20reducedDims%20slot.%20Dimensional%20reduction%20names%20are%20converted%20to%20upper%2Dcase%20(eg.%20%E2%80%9Cumap%E2%80%9D%20to%20%E2%80%9CUMAP%E2%80%9D)%20to%20match%20Monocle%203%20style) transfers the "umap" slot so that it's compatible with the "UMAP" slot required by monocle3.

Comment (gerbeldo) [reply r/R/trajectory_analysis.R]:
```suggestion
#' After conversion, this function also learns the trajectory graph.
```

**Review by gerbeldo — APPROVED** [review]
Some final nitpicky comments about docs, but after that LGTM 🎉

File: r/R/trajectory_analysis.R (hunk 31 above)
Comment (gerbeldo) [inline r/R/trajectory_analysis.R]:
```suggestion
runGenerateTrajectoryGraph <- function(req, data) {
```

File: r/R/trajectory_analysis.R [hunk 41]
```diff
@@ -0,0 +1,127 @@
+#' Generate UMAP and nodes coordinates for the initial trajectory analysis plot
+#'
+#' Returns a json object with UMAP and nodes coordinates, to be used for the initial plot (embedding + trajectory nodes)
+#'
+#' This represents the first step of the trajectory analysis. It allows the creation of the initial trajectory analysis plot, which is composed of an embedding and starting nodes (vertices in the lines).
+#' This plot is used to select the root nodes among the starting nodes. The root nodes will be then used in the following step of the trajectory analysis for pseudotime calculation.
+#'
+#' @param data SeuratObject
```
Comment (gerbeldo) [inline r/R/trajectory_analysis.R:14]:
```suggestion
#' @param req list of configuration parameters. Not used here.
#' @param data SeuratObject
```

File: r/R/trajectory_analysis.R (hunk 37 above)
Comment (saracastel) [reply r/R/trajectory_analysis.R]: that's right. I was getting confused because at some point I added monocle3::reduce_dimensions and then removed it

**Review by gerbeldo — APPROVED** [review]
